homeassistant/components/azure_data_explorer/__init__.py

```diff
@@ -0,0 +1,215 @@
+        schedules the first send.
+        """
+
+        self._listener_remover = self.hass.bus.async_listen(
+            MATCH_ALL, self.async_listen
+        )
+        self._schedule_next_send()
+
+    async def async_stop(self) -> None:
+        """Shut down the ADX by queueing None, calling send, join queue."""
+        if self._next_send_remover:
+            self._next_send_remover()
+        if self._listener_remover:
+            self._listener_remover()
+        self._shutdown = True
+        await self.async_send(None)
+
+    def update_options(self, new_options: dict[str, Any]) -> None:
+        """Update options."""
+        self._send_interval = new_options[CONF_SEND_INTERVAL]
+
+    async def test_connection(self) -> None:
+        """Test the connection to the Azure Data Explorer service."""
+        await self.hass.async_add_executor_job(self._client.test_connection)
+
+    def _schedule_next_send(self) -> None:
+        """Schedule the next send."""
+        if not self._shutdown:
+            if self._next_send_remover:
+                self._next_send_remover()
+            self._next_send_remover = async_call_later(
+                self.hass, self._send_interval, self.async_send
+            )
+
+    async def async_listen(self, event: Event) -> None:
+        """Listen for new messages on the bus and queue them for ADX."""
+        if state := event.data.get("new_state"):
+            await self._queue.put((2, (event.time_fired, state)))
+
+    async def async_send(self, _) -> None:
+        """Write preprocessed events to Azure Data Explorer."""
+
+        adx_events = []
+        dropped = 0
+        while not self._queue.empty():
+            _, event = self._queue.get_nowait()
+            adx_event, dropped = self._parse_event(*event, dropped)
+            self._queue.task_done()
+            if adx_event is not None:
+                adx_events.append(adx_event)
+
+        if dropped:
+            _LOGGER.warning(
+                "Dropped %d old events, consider filtering messages", dropped
+            )
+
+        if adx_events:
+            event_string = "".join(adx_events)
+
+            try:
+                await self.hass.async_add_executor_job(
+                    self._client.ingest_data, event_string
+                )
+
+            except KustoServiceError as err:
+                _LOGGER.error("Could not find database or table: %s", err)
+            except KustoAuthenticationError as err:
+                _LOGGER.error("Could not authenticate to Azure Data Explorer: %s", err)
+
+        self._schedule_next_send()
+
+    def _parse_event(
+        self,
+        time_fired: datetime,
+        state: State | None,
+        dropped: int,
+    ) -> tuple[str | None, int]:
+        """Parse event by checking if it needs to be sent, and format it."""
+
+        if state.state in FILTER_STATES or not self._entities_filter(state.entity_id):  # type: ignore[union-attr]
```

Comment:
If `state` is `None` the following `state.state` will raise an error. Please fix it accordingly

Reply:
I added the check, however I dont seem to be able to hit this code by any tests. so I think this is cought elsewhere in the code. My Python skills are not good enough to catch this.

Reply:
I fixed it.